### PR TITLE
Reporting table headers can now be configured separately for rows and columns

### LINKED ISSUE #14075: Independent setting for report Row/column headers (Document Formatting panel)
### Is your feature request related to a problem? Please describe.
Just a small inconvenience. Sometimes users want to read just row, or colum header, but not both at the same time.

### Describe the solution you'd like
NVDA should allow the user to configure if the user wants to read the header of colums, rows, both, or disabled.

### Describe alternatives you've considered
I checked the documentation to know if this can be done with some option.

### Additional context
In my opinion, the posibility to choose between read just colum or row headers, but not both aht the same time, can be useful when the user is reading a big table, for example in excel.

### PR DESCRIPTION
**Pull request by CyrilleB79**
### Link to issue number:
Closes #14075
### Summary of the issue:
It was asked in #14075 to be able to have table row/column headers reporting configured separately.

### Description of user facing changes
* A combo-box in Documentation formatting settings allows to select among 4 options to have table headers reported: Off, Columns and rows, Rows and Columns; it replaces the previous checkbox for table headers reporting.
* Setting row/column headers in Word or Excel (no UIA) is now possible even if the option to report table headers is disabled. Indeed, it is clearer from a UX point of view to allow it than to disallow it only when reporting table headers in the same direction (row/column) is disabled.
* The toggle script for table headers reporting has also been adapted to cycle through the 4 options.

### Description of development approach
* Implemented as per https://github.com/nvaccess/nvda/issues/14075#issuecomment-1240263166.
* Created an enumeration to store the possible options for this parameter, containing the numeric values for the config file and the display strings for the UI.
* Not directly linked but a a `noqa` exception for code complexity of `getObjectSpeech` has been removed because it was not relevant anymore. Excepted the removal of this #`noqa` directive, `getObjectSpeech` is not modified in this PR.

### Testing strategy
Manual testing of table navigation commands combining all the following possibilities:
* Tested in the following applications/controls:
  * Browsers (Chrome, Firefox, Edge, `ui.browseableMessage`)
  * Word with and without UIA
  * Excel without UIA (not possible to set table headers when using UIA, as described in #12882 )
  * Outlook message list
  * list view In task manager's tab "Details" (of processes)
* Tested with the 4 options: "Off", "Rows and columns", "Rows" and "Columns".
* Tested that headers were reported with speech with the appropriate options
* Tested that columns headers were reported with braille when "Rows and columns" or "Columns" option is enabled.

It has also been checked in `nvda.ini` that saving the config works as expected.

### Known issues with pull request:
None

### Change log entries:
New features (or Changes ?)
`Reporting table headers can now be configured separately for rows and columns. (#14075)`

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

File: source/config/configFlags.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# A part of NonVisual Desktop Access (NVDA)
+# Copyright (C) 2022 NV Access Limited, Cyrille Bougot
+# This file is covered by the GNU General Public License.
+# See the file COPYING for more details.
+
+"""Flags used for the configuration.
+"""
```
Comment (lukaszgo1) [inline source/config/configFlags.py:7]: [Python's conventions for docstrings](https://peps.python.org/pep-0257/) recommends to place closing triple quotes on the same line as the docstring if it is only a single line long.

File: source/config/configFlags.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+	
+	OFF = 0
+	ROWS_AND_COLUMNS = 1
+	ROWS = 2
+	COLUMNS = 3
+	
+	@property
+	def _displayStringLabels(self):
+		return {
+			# Translators: This is the label for a combobox in the
```
Comment (lukaszgo1) [inline source/config/configFlags.py:25]: The translator comment for these options seems wrong - these are not labels for the combo box.

File: source/config/profileUpgradeSteps.py [hunk 3]
```diff
@@ -118,3 +118,19 @@ def upgradeConfigFrom_6_to_7(profile: Dict[str, str]) -> None:
...
+	In Document formatting settings, "Row/column headers" check box has been replaced with "Headers" combobox.
+	"""
+	try:
+		reportTableHeaders = profile['documentFormatting']['reportTableHeaders']
+	except KeyError:
+		# Setting does not exist, no need for upgrade of this setting
+		log.debug("reportTableHeaders not present, no action taken.")
+	else:
+		if reportTableHeaders:
+			profile['documentFormatting']['reportTableHeaders'] = 1  # Rows and columns
```
Comment (lukaszgo1) [inline source/config/profileUpgradeSteps.py:134]: Can't you import the enum with possible values here and use `member.value` rather than an integer literal and a comment explaining what the value means?

File: source/globalCommands.py [hunk 4]
```diff
@@ -667,19 +667,17 @@ def script_toggleReportTables(self,gesture):
...
-		if config.conf["documentFormatting"]["reportTableHeaders"]:
-			# Translators: The message announced when toggling the report table row/column headers document formatting setting.
-			state = _("report table row and column headers off")
-			config.conf["documentFormatting"]["reportTableHeaders"]=False
-		else:
-			# Translators: The message announced when toggling the report table row/column headers document formatting setting.
-			state = _("report table row and column headers on")
-			config.conf["documentFormatting"]["reportTableHeaders"]=True
-		ui.message(state)
+		from config.configFlags import ReportTableHeaders
```
Comment (lukaszgo1) [inline source/globalCommands.py:674]: Why is this imported lazily?

Comment (CyrilleB79) [reply source/globalCommands.py:674]: It's imported lazily as done in more recent scripts of `globalCommands.py` such as `script_toggleScreenCurtain`. It seems relevant to avoid to have too much import statements at the top of this file when the imported variable will not be used elsewhere.

Let's wait for the opinion of NVAccess reviewer. I may put back the import at the top of the file if people prefer this way to do.

File: source/config/configFlags.py (hunk 1 above)
Comment (CyrilleB79) [reply source/config/configFlags.py]: Done in 450cfec36e9fca0c58713c7ce35ac32565296beb.

File: source/config/profileUpgradeSteps.py (hunk 3 above)
Comment (CyrilleB79) [reply source/config/profileUpgradeSteps.py]: Done in 450cfec36e9fca0c58713c7ce35ac32565296beb.

**Comment by seanbudd** [comment]
I think a more intuitive UX for the option would be checkboxes:

- [ ] Report table headers (checking this checks both, if one option is checked it is indeterminate)
  - [ ] Rows
  - [ ] Columns
  
What do you think?

**Comment by seanbudd** [comment]
Sorry disregard my previous comment, missed the suggestion in the issue

File: source/speech/speech.py [hunk 5]
```diff
@@ -611,12 +612,7 @@ def speakObject(
 		speak(sequence, priority=priority)
 
 
-
-
-# C901 'getObjectSpeech' is too complex
-# Note: when working on getObjectSpeech, look for opportunities to simplify
-# and move logic out into smaller helper functions.
-def getObjectSpeech(  # noqa: C901
```
Comment (seanbudd) [inline source/speech/speech.py:619]: why was this removed?

Comment (seanbudd) [reply source/speech/speech.py:619]: in fact, where possible, please factor out the new changes into helper function(s)

**Review by seanbudd — COMMENTED** [review]
Thanks @CyrilleB79, generally looks good.

File: source/speech/speech.py (hunk 5 above)
Comment (CyrilleB79) [reply source/speech/speech.py:619]: @seanbudd wrote:
> why was this removed?

As written in initial description
> Not directly linked but a a noqa exception for code complexity has been removed because it was not useful anymore.

To be more specific, I have not modified `getObjectSpeech` in this PR but had to reread it while developing. I have figured that `getObjectSpeech` is not complex anymore and thus have removed this `#noqa` directive that was already useless before this PR.

@seanbudd wrote:
> in fact, where possible, please factor out the new changes into helper function(s)

Where do you suggest me to factor code?

Comment (seanbudd) [reply source/speech/speech.py:619]: I was hoping to confirm that the function is under the limits.

> >in fact, where possible, please factor out the new changes into helper function(s)
> 
> Where do you suggest me to factor code?

apologies, I thought the below changes were in the function

**Review by seanbudd — DISMISSED** [review]

File: source/speech/speech.py (hunk 5 above)
Comment (CyrilleB79) [reply source/speech/speech.py:619]: I have updated initial description to clarify this point.

File: source/globalCommands.py (hunk 4 above)
Comment (CyrilleB79) [reply source/globalCommands.py:674]: @seanbudd what is your opinion on this lazy import?

Comment (seanbudd) [reply source/globalCommands.py:674]: Since config is already imported at the top level, this should probably be moved out.
I believe the screen curtain case (and others) is more about avoid circular dependencies.

**Review by Qchristensen — APPROVED** [review]
User Guide change looks fine, good work.

**Comment by leonardder** [comment]
Please don't merge yet as I have some suggestions I will post later today.

File: source/NVDAObjects/IAccessible/winword.py [hunk 6]
```diff
@@ -205,7 +206,7 @@ def fetchAssociatedHeaderCellText(self,cell,columnHeader=False):
 
 	def script_setColumnHeader(self,gesture):
 		scriptCount=scriptHandler.getLastScriptRepeatCount()
-		if not config.conf['documentFormatting']['reportTableHeaders']:
+		if config.conf['documentFormatting']['reportTableHeaders'] == ReportTableHeaders.OFF:
```
[the remark this answers is not in the document]
Comment (CyrilleB79) [reply source/NVDAObjects/IAccessible/winword.py:209]: Actually, setting or removing column headers while only row headers are reported is working. If you remove this "if" statement, i.e. allow this script to work regardless of this reportTableHeaders configuration, it works too.

For simplicity (regarding the error message), I am thinking to remove this if statement, i.e. setting column headers even when reporting table headers is disabled. @leonardder, @seanbudd, Can you see any drawback to this solution?

[the remark this answers is not in the document]
Comment (CyrilleB79) [reply source/NVDAObjects/IAccessible/winword.py]: I have finally allowed setting/removing row/column headers even if reporting table headers is disabled. I have updated the initial description accordingly.
@seanbudd, @leonardder, let me know if this is OK for you.

[the remark this answers is not in the document]
Comment (leonardder) [reply source/NVDAObjects/IAccessible/winword.py]: Great!

**Review by LeonarddeR — APPROVED** [review]

**Review by seanbudd — APPROVED** [review]
LGTM, thanks @CyrilleB79

**Comment by CyrilleB79** [comment]
@seanbudd, I just realize that this PR was an API breaking change, due to the configuration modification and upgrade, but it has been merged in 2022.4 which was not intended to be an API breaking release.

I don't know of an add-on that uses this config keys.

This has no sense to revert it now. However, should this API breaking change be advertised somewhere? I would say at least on the API mailing list. And maybe in the change log, even if usually the log of older releases is not amended.

**Comment by josephsl** [comment]
Hi, StationPlaylist add-on uses this config key, but as of now the add-on code will work on both 2022.3 and 2022.4 as integer values above 0 are seen as “true” in Python. I do plan to edit the add-on code son to use 2022.4 changes anyway. Thanks.

**Comment by paulber19** [comment]
Hi,
I would like to know how an API break is defined.
It seems to me that any modification of NVDA's code is an API break for 
add-ons that rely on it.
For example, documentBase.py has undergone a major overhaul in version 
2022.2 compared to 2022.1 to add in particular the commands for moving 
tables in tables.
And that involved a serious upgrade in one of my add-ons.
  So it seems to me that for developers of add-ons embedded in NVDA, any 
new version of NVDA is an API break and it is necessary to 
systematically check that the NVDA code on which it was based did not 
not undergone any modification.
That's why I don't quite understand the principle of API breaking only 
when releasing the first version of the year.
Best regards.

**Comment by josephsl** [comment]
Hi, it didn’t affect StationPlaylist ad-on, so all is good with me. Thanks.

**Comment by paulber19** [comment]
Really sorry for intervening in this thread which was about API breaking.
All is well, forget my intervention.
Best regards.
Paul.

**Comment by seanbudd** [comment]
Hi Cyrille,
I plan to announce this on the API mailing list.

Is this an accurate summary of the API breakage?

```
The configuration specification has been altered in the [documentFormatting] section.
The value of reportTableHeaders is now a member of configFlags.ReportTableHeaders, instead of a boolean.  (#14132)
 ```
